nvflare/app_common/executors/ham.py

```diff
@@ -176,76 +212,323 @@ def execute(self, task_name: str, shareable: Shareable, fl_ctx: FLContext, abort
+            self._num_aggrs = 0
         except Exception as ex:
-            self.log_error(fl_ctx, f"exception 'aggregate' from {type(self.aggregator)}: {secure_format_exception(ex)}")
-            return make_reply(ReturnCode.EXECUTION_EXCEPTION)
+            self.log_error(
+                fl_ctx, f"exception 'aggregate' from {type(self._aggregator)}: {secure_format_exception(ex)}"
+            )
+            self._process_error = True
+            result = None
+        return result
+
+    def _process_task_end(self, topic: str, request: Shareable, fl_ctx: FLContext) -> Shareable:
+        # process notification from parent that task is ended
+        task_seq = request.get_header(EdgeTaskHeaderKey.TASK_SEQ)
+
+        if self._num_children > 0:
+            # fire-and-forget notification to all my children
+            req = Shareable()
+            req.set_header(EdgeTaskHeaderKey.TASK_SEQ, task_seq)
+            engine = fl_ctx.get_engine()
+            engine.send_aux_request(
+                targets=self._children,
+                topic=topic,
+                request=req,
+                timeout=0,  # fire and forget
+                fl_ctx=fl_ctx,
+                optional=True,
+            )
+
+        task_info = self._pending_task
+        if task_info:
+            if task_info.seq <= task_seq:
+                # my current task is before the ended task - end my task
+                self.log_info(
+                    fl_ctx, f"ended current task seq {task_info.seq}: got end_task from parent for task {task_seq}"
+                )
+                self._task_done = True
+            else:
+                self.log_info(
+                    fl_ctx, f"ignored end_task from parent for task {task_seq} since it's < my task seq {task_info.seq}"
+                )
+        else:
+            self.log_info(fl_ctx, f"ignored end_task from parent for task {task_seq} since I have no current task")
+
+        return make_reply(ReturnCode.OK)
+
+    def _process_aggr_result(self, topic: str, request: Shareable, fl_ctx: FLContext) -> Shareable:
+        self.log_info(fl_ctx, f"processing aggregation result report: {topic}")
+
+        task_info = self._pending_task
+        if task_info:
+            assert isinstance(task_info, TaskInfo)
+            seq = task_info.seq
+            current_round = task_info.round
+        else:
+            seq = 0
+            current_round = -1
+
+        accepted, reply = process_aggr_result_from_child(
+            processor=self,
+            request=request,
+            current_task_seq=seq,
+            fl_ctx=fl_ctx,
+            accept_f=self._accept_result,
+            current_round=current_round,
+        )
+
+        peer_ctx = fl_ctx.get_peer_context()
+        assert isinstance(peer_ctx, FLContext)
+        client_name = peer_ctx.get_identity_name()
+        self.log_info(fl_ctx, f"processed aggr result report from {client_name} at round {current_round}: {accepted=}")
```

Comment:
Turn it into debug?

Reply:
Sure we can change it once this whole thing is proven.